File: backend/app/api/artists.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query

from app.db import get_driver
from app.models.schemas import ArtistDetail, ArtistSummary
from app.services import queries

router = APIRouter(prefix="/artists", tags=["artists"])


@router.get("", response_model=list[ArtistSummary])
def artists(
    q: str | None = None,
    genre: str | None = None,
    generation: int | None = Query(None, ge=1, le=6),
) -> list[ArtistSummary]:
    return queries.list_artists(get_driver(), q=q, genre=genre, generation=generation)


@router.get("/hubs", response_model=list[ArtistSummary])
def hubs(limit: int = Query(12, ge=1, le=50)) -> list[ArtistSummary]:
    return queries.list_hubs(get_driver(), limit=limit)


@router.get("/{artist_id}", response_model=ArtistDetail)
def artist(artist_id: str) -> ArtistDetail:
    detail = queries.get_artist(get_driver(), artist_id)
    if not detail:
        raise HTTPException(404, "Artist not found")
    return detail

File: frontend/src/lib/api/artists.ts
import { apiGet } from "./client";
import type { ArtistSummary, ArtistDetail } from "@/types/artist";

export const artists = (q?: string, genre?: string, generation?: number) => {
  const params = new URLSearchParams();
  if (q) params.set("q", q);
  if (genre) params.set("genre", genre);
  if (generation) params.set("generation", String(generation));
  const qs = params.toString();
  return apiGet<ArtistSummary[]>(`/api/artists${qs ? `?${qs}` : ""}`);
};

export const hubs = (limit = 12) => apiGet<ArtistSummary[]>(`/api/artists/hubs?limit=${limit}`);

export const artist = (id: string) => apiGet<ArtistDetail>(`/api/artists/${id}`);
